### ISSUE TITLE
[FR] Should be possible to make the guest app only available for certain user groups

### ISSUE THREAD

**Issue by haukman** [owncloud/guests#92]
As an ownCloud admin you don´t want every user to be able to create a guest user - therefore it would be good to restrict the usage of this app only to certain groups (comparable to the market app).

**Comment by cortho** [comment]
This issue occurred to me as well. Since the guests app is of type _authentication_ (info.xml) it is excluded from making it possible to only provide for selected groups.

I filed an issue in core two weeks ago https://github.com/owncloud/core/issues/27995

**Comment by cortho** [comment]
closing https://github.com/owncloud/guests/issues/90 (because of the better headline here).

Issue was:

In info.xml the guests app is set to type authentication. With this setting, it is not possible, to enable/disable this app only for specific groups, as the type 'authentication' is one of the types (filesystem, prelogin, authentication, logging, prevent_group_restriction) which are except in https://github.com/owncloud/core/blob/master/settings/js/admin-apps.js#L171 ff

TL;DR with this app type it is not possible to allow this app only for specific groups

**Comment by pmaier1** [comment]
> TL;DR with this app type it is not possible to allow this app only for specific groups

What does it need to enable limiting the app to certain groups? Can we simply change the app type or does this have other implications? Maybe also a question for @PVince81

**Comment by cortho** [comment]
The basic question for me is, why an app needs to be of such a type in order to be able to use hooks. I don't see a required relation here, so I filed the issue https://github.com/owncloud/core/issues/27995 where I didn't really get a reply.

With the current implementation we send the initial invitation mail by using the post_shared hook. This, howerver is worth having a look at too, maybe there is a better solution.

Most probably we should find out how to send invitation email in another way and focus on this. 
(The issue in core still is relevant, as I'd like to use this for another app as well)

**Comment by PVince81** [comment]
For any situation where "disabling an app for groups" is to be considered, need to think about these:
- what happens if the app's code/backend is not loaded whenever filesystem/sharing operations are in place ?
- does the app provide user/group backends ?

In both these cases, having the app not loaded for specific users will likely cause side effects.
If there is code that require for example accessing a user/group from the backend provided by the app (for whatever reason), it will not see the custom user/groups at all.

**Comment by pmaier1** [comment]
Not sure if I fully understand the issue yet. Probably at least one of those would apply to the guests app, I guess. In any case we need to find an ability to restrict guest user creation to certain groups. Sharing with existing guests should work for all users in the contrary but this should work anyway as those are regular users in the end. This has very high importance for the adoption of the guests app in enterprise scenarios.

**Comment by pmaier1** [comment]
To make things a little more clear here:
It is not needed here to restrict the whole guests app to certain groups. Guest user **creation** (implying UI elements and backend) needs to be restrictable for certain groups. Probably this is a feature the guests app needs to provide itself.

Please comment on feasibility and efforts, thanks!

**Comment by cortho** [comment]
I'll have to take some time to look deeper into the code having the mentioned prerequisites in mind, I'll share my findings then.

**Comment by PVince81** [comment]
> Probably this is a feature the guests app needs to provide itself.

In that case, yes.

@cortho for any dev chat / question feel free to ping me on Freenode in #owncloud-dev (PVince81)

**Comment by PVince81** [comment]
See my comment here: https://github.com/owncloud/core/issues/27995#issuecomment-318658403

I don't think you can use "Enable app for groups" core feature as it would disable important hooks.

**Comment by PVince81** [comment]
release blocker for the current guests release ? (p1)

**Comment by PVince81** [comment]
Since we can't use the "Enable app for groups" due to backend concerns, here's another idea:
- add a new field in the guest app "only enable guest creation for the following groups:" with an autocomplete where the admin can select any group.
    - if no group is selected, any user can create guests
    - if one or multiple groups are specified, only users in these groups are able to create guests
    - for any user not in the selected groups:
        - don't load the guests JS code in the files app, would disable the UI part
        - adjust the backend to also refuse creating guest users in case someone tries to hack it

See https://github.com/owncloud/impersonate/blob/master/templates/settings-admin.php#L10 for an example how we did something similar for the impersonate app. https://github.com/owncloud/impersonate/blob/master/js/settings-admin.js#L2

**Comment by pmaier1** [comment]
Thanks a lot for your input! As this ticket is in milestone 'development' it's a very important item for the release after the current one. Current release is in milestone 'QA' and just needs to be pushed out.

**Comment by pmaier1** [comment]
Cleaning up current milestone as we need a release soon -> no development started here, moving this to the next version.

**Comment by pmaier1** [comment]
@PVince81 can you give a very rough estimate for https://github.com/owncloud/guests/issues/92#issuecomment-326248682 ?
